### PR TITLE
Reader Refresh: Adding followSource to figure out where follows come from

### PR DESCRIPTION
**Pull request by samouri**
This PR aims to enable us to be able to differentiate between:
1. search result vs. search recommendations 

fixes: #9883

**Current `source` property for follows track events:**
There is a function within `client/reader/stats.js` called `recordFollow`.  It relies upon [the current url](https://github.com/Automattic/wp-calypso/blob/c8f93e9c007245f7a6018d6245bd522d32ae6a8a/client/reader/stats.js#L42) in order to figure out the follow-_source_.  This has been extremely nice until now because we never needed to differentiate between follows on the same page and it allowed us to essentially forgo the ceremony of passing args through a tower of components.

**I see two reasonable options forward:**
1. a new abstraction: Add a new prop to `post` for `postType` (at least to differentiate recs from reg posts but maybe could be extended further).  Since we are already passing post everywhere the plumbing work is already done.
2. pass through `additionalProps` through the entire component tree (e.g. from SearchStream --> individual FollowButtons) without any additional smarts.  

For now I'm leaning toward number 2 because even though it is a lot more boilerplate and feels _dumber_, it is technically more flexible and will give us room to easily refactor in the future when we have a greater understanding of how we want to vary posts within streams ( and the different analytics needs we will have)

@blowery @bluefuton  thoughts?

### CODE REVIEW

**Comment by bluefuton** [comment]
Looks like a solid approach @samouri 👍 

How about adding constants for the different follow sources rather than using string values? They could live inside a new `blocks/follow-button/constants.js` perhaps.

**Comment by samouri** [comment]
@bluefuton: done

File: client/reader/search-stream/index.jsx [hunk 1]
```diff
@@ -27,6 +27,10 @@ import Suggestion from './suggestion';
 import ReaderPostCard from 'blocks/reader-post-card';
 import { RelatedPostCard } from 'blocks/reader-related-card-v2';
 import config from 'config';
+import {
+	EMPTY_SEARCH_RECOMMENDATIONS,
+	SEARCH_RESULTS,
+ } from 'reader/follow-button/constants';
```
Comment (bluefuton) [inline client/reader/search-stream/index.jsx]: Sorry @samouri - poor filename suggestion on my part. Shall we call the file `follow-sources.js` instead? (Consistent with `action-types.js` in Redux then.)

Comment (samouri) [inline client/reader/search-stream/index.jsx]: i have no objections 👍

**Comment by bluefuton** [comment]
@samouri I think we should send `follow_source` with the Tracks event rather than `followSource`, to match the other params:

screen shot 2016-12-15 at 15 36 49

Could use lodash `snakeCase` for this.

**Comment by samouri** [comment]
Fixed merge conflicts and addressed the comment by @bluefuton

**Comment by blowery** [comment]
I actually don't like having the values being spec'd in `follow-button/sources`. You're teaching follow button about all the places it might be used, and the values are only used once in the app. I'd rather have a constant at the point of use, not up in the block.

**Comment by samouri** [comment]
@blowery I think I disagree here. The value added by having them in `follow-sources.jsx` isn't in having the follow button know or care about the constants.  Instead its to help us write self-documenting code.  

The effort required in figuring out: 'what are the potential values for `follow_source`'? goes way down by having this structure.  
Separately I believe we should do this for all of our tracks event names.

**Comment by blowery** [comment]
> Separately I believe we should do this for all of our tracks event names. (reader/stats/event-names.js)

I'm happy they're in constants, and I'd be fine with that. I just don't want the constants in `blocks` or `components`. They're Reader specific, so they should live in `reader` somewhere.

**Comment by samouri** [comment]
@blowery agreed.
I think its currently in `client/reader/follow-button/follow-sources.jsx`.  Does that count as in `reader` or do you mean that it should exist somewhere like in `client/reader/stats/constants.js`

**Comment by bluefuton** [comment]
@samouri I think they're alright in `client/reader/follow-button/follow-sources.jsx` 👍

**Comment by blowery** [comment]
Agreed, that's fine.

**Comment by samouri** [comment]
@bluefuton: I've been hesitating to merge this because I'm not too happy with the solution.

Do you think if we included which endpoint the post object came from within the post object we could create a cleaner solution?  I think there used to be a 1:1 relationship between `location` (as gotten from `getLocation` and the source of the post.  But now that we have things like in-stream-recs and also recs on the search page, this is no longer true.  If anything I think we are headed further in the direction of more variety within a single page.

Another issue this would help solve is https://github.com/Automattic/wp-calypso/issues/10166.  Having the source endpoint at normalization time would help us enable rules to behave differently for posts from different sources.

**Comment by bluefuton** [comment]
> Do you think if we included which endpoint the post object came from within the post object we could create a cleaner solution?

Could there be a scenario where the same post comes from multiple endpoints though? Recommended by two different algorithms, or loaded for a tag stream but also appears in an in-stream recommendation?

**Comment by samouri** [comment]
good point @bluefuton.  I'll rebase + merge with current approach

**Pull request merged**